src/Mvc/Mvc.ViewFeatures/src/Infrastructure/DefaultTempDataSerializer.cs

```diff
@@ -46,19 +45,17 @@ internal class DefaultTempDataSerializer : TempDataSerializer
                         break;
 
                     case JsonValueType.String:
-                        var stringValue = item.Value.GetString();
-                        // BsonTempDataSerializer will parse certain types of string values. We'll attempt to imitiate it.
-                        if (DateTime.TryParseExact(stringValue, "r", CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out var dateTime))
+                        if (item.Value.TryGetGuid(out var guid))
                         {
-                            deserializedValue = dateTime;
+                            deserializedValue = guid;
                         }
-                        else if (Guid.TryParseExact(stringValue, "B", out var guid))
+                        else if (item.Value.TryGetDateTime(out var dateTime))
                         {
-                            deserializedValue = guid;
+                            deserializedValue = dateTime;
                         }
                         else
                         {
-                            deserializedValue = stringValue;
+                            deserializedValue = item.Value.GetString();
```

Comment:
Will the JSON strings more often than not be GUID/DateTimes?

Reply:
I'd expect the latter. It's not uncommon for users to JSON serialize on complex models so that they could put it in this dictionary. Is there a way to improve this?